kube-core/src/labels.rs

```diff
@@ -0,0 +1,625 @@
+#![allow(missing_docs)]
+use core::fmt;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{LabelSelector, LabelSelectorRequirement};
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::PartialEq,
+    collections::{BTreeMap, BTreeSet},
+    fmt::Display,
+    iter::FromIterator,
+    option::IntoIter,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error("failed to parse value as expression: {0}")]
+/// Indicates failure of conversion to Expression
+pub struct ParseExpressionError(pub String);
+
+// local type aliases
+type Expressions = Vec<Expression>;
+
+/// Matcher trait for implementing alternalive Selectors
+pub trait Matcher {
+    type Items;
```

Comment:
Given that we are not sure exactly how to take this, and we don't actually have `impl Matcher` in any signatures anymore, we can probably seal and rename this trait and so we are free to change it. It shouldn't block anything upstream, because it's just a user convenience at this point (and it is a nice convenience).

```suggestion
/// Selector extension trait for querying selector like objects
///
/// Only implemented by `Selector` and `Expression`.
pub trait SelectorExt: private::Sealed {
    /// Collection type to compare with self
    type Search
```

(also think it makes sense to keep this as an associated type because then you can in theory do matches from different types i.e. you could do `labels.matches(btreemap)` and `labels.matches(hashmap)` or maybe `labels.matches(iterator_of_expressions)`)

Reply:
Yeah, that would work and it can be expanded later. I feel like this is now in the good shape. WDYT?

Reply:
Looks great!